# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [26]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('ICCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
598          599         0       3   
127          128         1       3   
535          536         1       2   
617          618         0       3   
878          879         0       3   
103          104         0       3   
620          621         0       3   
390          391         1       1   
655          656         0       2   
525          526         0       3   

                                                Name     Sex   Age  SibSp  \
598                                Boulos, Mr. Hanna    male   NaN      0   
127                        Madsen, Mr. Fridtjof Arne    male  24.0      0   
535                           Hart, Miss. Eva Miriam  female   7.0      0   
617  Lobb, Mrs. William Arthur (Cordelia K Stanlick)  female  26.0      1   
878                               Laleff, Mr. Kristo    male   NaN      0   
103                       Johansson, Mr. Gustaf Joel    male  33.0      0   
620                              Yasbeck, Mr. Antoni    male  27.0      1   
390                       Carter, Mr. William Ernest    male  36.0      1   
655                        Hickman, Mr. Leonard Mark    male  24.0      2   
525                               Farrell, Mr. James    male  40.5      0   

     Parch        Ticket      Fare    Cabin Embarked  
598      0          2664    7.2250      NaN        C  
127      0       C 17369    7.1417      NaN        S  
535      2  F.C.C. 13529   26.2500      NaN        S  
617      0     A/5. 3336   16.1000      NaN        S  
878      0        349217    7.8958      NaN        S  
103      0          7540    8.6542      NaN        S  
620      0          2659   14.4542      NaN        C  
390      2        113760  120.0000  B96 B98        S  
655      0  S.O.C. 14879   73.5000      NaN        S  
525      0        367232    7.7500      NaN        Q

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [2]:
passengers['Survived'].sample(10)

697    1
653    1
238    0
288    1
649    1
284    0
265    0
796    1
153    0
684    0
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [3]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [4]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [12]:
# todo write code here
passengers['Sex'].value_counts()
passengers['Survived'].value_counts()
passengers[['Sex','Survived'] ]

Sex  Survived
0      male         0
1    female         1
2    female         1
3    female         1
4      male         0
5      male         0
6      male         0
7      male         0
8    female         1
9    female         1
10   female         1
11   female         1
12     male         0
13     male         0
14   female         0
15   female         1
16     male         0
17     male         1
18   female         0
19   female         1
20     male         0
21     male         1
22   female         1
23     male         1
24   female         0
25   female         1
26     male         0
27     male         0
28   female         1
29     male         0
..      ...       ...
861    male         0
862  female         1
863  female         0
864    male         0
865  female         1
866  female         1
867    male         0
868    male         0
869    male         1
870    male         0
871  female         1
872    male         0
873    male         0
874  female         1
875  female         1
876    male         0
877    male         0
878    male         0
879  female         1
880  female         1
881    male         0
882  female         0
883    male         0
884    male         0
885  female         0
886    male         0
887  female         1
888  female         0
889    male         1
890    male         0

[891 rows x 2 columns]

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [13]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [14]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [15]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [16]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [17]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass                                   Name  \
877          878         0       3                   Petroff, Mr. Nedelio   
169          170         0       3                          Ling, Mr. Lee   
555          556         0       1                     Wright, Mr. George   
246          247         0       3  Lindahl, Miss. Agda Thorilda Viktoria   
333          334         0       3        Vander Planke, Mr. Leo Edmondus   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked AgeCat  
877    male  19.0      0      0  349212   7.8958   NaN        S  Adult  
169    male  28.0      0      0    1601  56.4958   NaN        S  Adult  
555    male  62.0      0      0  113807  26.5500   NaN        S  Adult  
246  female  25.0      0      0  347071   7.7750   NaN        S  Adult  
333    male  16.0      2      0  345764  18.0000   NaN        S  Child

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [19]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [20]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [21]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [24]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 
agecat_comparision_df = pd.DataFrame({ 'AllPassengers' : agecat_all_series, 'Survivors' : agecat_survivor_series })
agecat_comparision_df['AgeSurvivialRate'] = agecat_comparision_df['Survivors'] / agecat_comparision_df['AllPassengers']
agecat_comparision_df

AllPassengers  Survivors  AgeSurvivialRate
Adult            575        220          0.382609
Child            139         70          0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [29]:
# todo: repeat the analysis in the previous cell for Pclass 
pclass_all_series = passengers['Pclass'].value_counts()
pclass_survivor_series = survivors['Pclass'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 
pclass_comparision_df = pd.DataFrame({ 'AllPassengers' : pclass_all_series, 'Survivors' : pclass_survivor_series })
pclass_comparision_df['PassengerClassSurvivialRate'] = pclass_comparision_df['Survivors'] / pclass_comparision_df['AllPassengers']
pclass_comparision_df


AllPassengers  Survivors  PassengerClassSurvivialRate
1            216        136                     0.629630
2            184         87                     0.472826
3            491        119                     0.242363